blogposts/2022/introducting-migrator-service.md

```diff
@@ -0,0 +1,120 @@
+
+- **Phase 1**: Create new catalog structures for the index, but do not begin to populate them yet. Transactions that started before the completion of this phase will only see (and therefore only write to) the old catalog structures.
+- **Phase 2**: Wait for all transactions starting before the end of **Phase 1** to finish. Then, perform a full-table scan and write the content of each row to the new index structure. Open the new index structure up to writes.
+- **Phase 3**: Wait for all transactions starting before the end of **Phase 2** to finish. Then, perform another full-table scan and write the content of each row inserted during **Phase 2** to the new index structure. The old index structure can now be closed to writes.
+
+There's a particular detail here that deserves special attention in hindsight: concurrent index creation operations block until **all transactions started before them** have either committed or rolled back. Mixed with long-held advisory locks, we encounter this rather nasty situation.
+
+In `psql` **Process A**, we acquire an advisory lock.
+
+```sql
+[A] sourcegraph@sourcegraph ❯ SELECT pg_advisory_lock(0, 0);
+ pg_advisory_lock
+------------------
+
+(1 row)
+
+Time: 20.616 ms
+```
+
+In `psql` **Process B**, we try to acquire an advisory lock but block as it's already held by another process.
+
+```sql
+[B] sourcegraph@sourcegraph ❯ SELECT pg_advisory_lock(0, 0);
+
+```
+
+Back in `psql` **Process A**, we attempt to create an index concurrently on _literally any table_. After a second or so, the Postgres deadlock detector kicks in and cancels the concurrent index creation.
+
+```sql
+[A] sourcegraph@sourcegraph ❯ CREATE INDEX CONCURRENTLY test ON repo(name, deleted_at);
+ERROR:  40P01: deadlock detected
+DETAIL:  Process 2149 waits for ShareLock on virtual transaction 3/13053932; blocked by process 2471.
+Process 2471 waits for ExclusiveLock on advisory lock [16384,0,0,2]; blocked by process 2149.
+HINT:  See server log for query details.
+LOCATION:  DeadLockReport, deadlock.c:1153
+Time: 1107.901 ms (00:01.108)
+```
+
+This error is a bit dense, but states that **Process A** and **Process B** are holding locks the other one wants. The concurrent index creation is waiting for **Process B**'s attempt to acquire the advisory lock to finish, but is blocked on **Process A**'s ownership of that lock.
+
+Additionally, this leaves the database with a _partially created_ index, which exists in the system catalog but is available neither for reads nor writes. To resolve this database state previously, an administrator must drop and recreate the index manually. This is an easy step to forget, and one that can have very severe performance penalties. In newer versions of the migrator, invalid indexes are automatically detected and repaired.
+
+TODO - need more content here
```

Comment:
- [ ] Tie-in back to original error and add a more exciting introduction to the second half of the solution